azurelinuxagent/ga/rsm_version_updater.py

```diff
@@ -0,0 +1,159 @@
+from azurelinuxagent.common.exception import AgentUpgradeExitException, AgentUpdateError
+from azurelinuxagent.common.utils.flexible_version import FlexibleVersion
+from azurelinuxagent.common.version import CURRENT_VERSION, AGENT_NAME
+from azurelinuxagent.ga.ga_version_updater import GAVersionUpdater, VMDisabledRSMUpdates
+from azurelinuxagent.ga.guestagent import GuestAgent
+
+
+class RSMVersionUpdater(GAVersionUpdater):
+    def __init__(self, gs_id, daemon_version, last_attempted_rsm_version_update_time):
+        super(RSMVersionUpdater, self).__init__(gs_id)
+        self._daemon_version = daemon_version
+        self._last_attempted_rsm_version_update_time = last_attempted_rsm_version_update_time
+
+    @staticmethod
+    def _get_all_agents_on_disk():
+        path = os.path.join(conf.get_lib_dir(), "{0}-*".format(AGENT_NAME))
+        return [GuestAgent.from_installed_agent(path=agent_dir) for agent_dir in glob.iglob(path) if
+                os.path.isdir(agent_dir)]
+
+    def _get_available_agents_on_disk(self):
+        available_agents = [agent for agent in self._get_all_agents_on_disk() if agent.is_available]
+        return sorted(available_agents, key=lambda agent: agent.version, reverse=True)
+
+    def is_update_allowed_this_time(self):
+        """
+        update is allowed once per (as specified in the conf.get_autoupdate_frequency())
+        If update allowed, we update the last_attempted_rsm_version_update_time to current time.
+        """
+        now = datetime.datetime.now()
+
+        if self._last_attempted_rsm_version_update_time != datetime.datetime.min:
+            next_attempt_time = self._last_attempted_rsm_version_update_time + datetime.timedelta(
+                seconds=conf.get_agentupdate_frequency())
+        else:
+            next_attempt_time = now
+
+        if next_attempt_time > now:
+            return False
+        self._last_attempted_rsm_version_update_time = now
+        # The time limit elapsed for us to allow updates.
+        return True
+
+    def check_and_switch_updater_if_changed(self, agent_family, gs_id):
+        """
+        Checks if there is a new goal state and decide if we need to continue with rsm update or switch to self-update.
+        Firstly it checks agent supports GA versioning or not. If not, we raise exception to switch to self-update.
+        if vm is enabled for RSM updates and continue with rsm update, otherwise we raise exception to switch to self-update.
+        if either isVersionFromRSM or isVMEnabledForRSMUpgrades is missing in the goal state, we ignore the update as we consider it as invalid goal state.
+        """
+        if self._gs_id != gs_id:
+            self._gs_id = gs_id
+            if not conf.get_enable_ga_versioning():
+                raise VMDisabledRSMUpdates()
+
+            if agent_family.is_vm_enabled_for_rsm_upgrades is None:
+                raise AgentUpdateError(
```

Comment:
if we have a VM which was enabled for RSM upgrades and then is switched to a host which does not send down the is_vm_enabled_for_rsm_upgrades, then we will not process any goal states?

What is the solution for customers who may end up getting blocked by this? I'm thinking we can use the GA versioning flag in conf. Is there another way?

Reply:
As we discussed offline, there isn't much we can do about it.